server-common/src/main/java/org/apache/kafka/server/common/FeatureVersion.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.common;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public interface FeatureVersion {
+
+    /**
+     * The level of the feature. 0 means the feature is disabled.
+     */
+    short featureLevel();
+
+    /**
+     * The name of the feature.
+     */
+    String featureName();
+
+    /**
+     * The next metadata version to be released when the feature became production ready.
+     * (Ie, if the current production MV is 17 when a feature is released, its mapping should be to MV 18)
+     */
+    MetadataVersion metadataVersionMapping();
```

Comment:
I am still confused by this one. Based on our offline discussion, my understanding is that this is only used during bootstrapping. We should try to make this clear in the name and in the javadoc.

> (Ie, if the current production MV is 17 when a feature is released, its mapping should be to MV 18)

For my understanding, why do we require to be the next one? Requiring the current seems more natural but I may be missing something.

Reply:
I would like the mapping to remain consistent for MV. If we do the current mapping, some versions/images with MV X will have feature version Y and some will not (since it wasn't created when MV X was first released)

Sorry I forgot to change the name, I will do that today 👍

Reply:
> I would like the mapping to remain consistent for MV. If we do the current mapping, some versions/images with MV X will have feature version Y and some will not (since it wasn't created when MV X was first released)

Isn't it something that will happen anyway? For instance, I will add `group.version=1` soon and it will require an old MV. Likely the oldest one supported by kraft.

Reply:
Hmm, there are a lot of components that are all getting mixed around here. 

1. All features require IBP-3.3.0-IV0. This is because this is the minimum MV to bootstrap and write a feature record :)  This is encoded in Features#validateVersion and doesn't need to be specified for all features since it will be required by all.
2. Some features may require a specific other feature version in order to be set. None of the features we proposed (transaction/group coordinator) need this but it was requested in the KIP, so I have implemented a framework to do so.
3. We need to set a reasonable default feature for when folks bootstrap using only metadata. There are a few options here. One is to set all features to 0. We ended up deciding to take the latest features as default. There is a small wrinkle here in that for a given MV, we may introduce a feature version after the code that releases the MV. If folks are running off trunk, they will have different features for the same MV if we choose the current MV, and that's why I suggest the next one. 

This bootstrap method is only used for 3. 1 and 2 are covered by the validateVersion method and by enumerating non 3.3 dependencies when defining the featureVersion.